[CRDB-15892] SQL: NULL geography concatenation is inconsistent with Postgres

surahman opened the pull request:
Github Issue [#79846](https://github.com/cockroachdb/cockroach/issues/79846)
JIRA ticket [CRDB-15892](https://cockroachlabs.atlassian.net/browse/CRDB-15892)


- [x] `Delimiter` in `sql/types/`:  `Box2D` `;` , for `Geometry`  and Geography` `:`, and for everything else `,`
- [x] Change the delimiter for `Format` for [`DArray`](https://github.com/cockroachdb/cockroach/blob/b9a675266303528d71b388ac35dfb87f145cc7a8/pkg/sql/sem/tree/datum.go#L4414)
- [x] Change `pg_type` table to return the correct `typdelim` in [`pgCatalogTypeTable`](https://github.com/cockroachdb/cockroach/blob/4c32814a93a9447abcc12c168a3a45a865cd3785/pkg/sql/pg_catalog.go#L2938)
- [ ] Add the relevant logic tests.
- [ ] Fix broken tests in `TestRandParseDatumStringAs` and `TestAvroSchema`.
- [ ] Add [test](https://github.com/cockroachdb/cockroach/blob/1a707a1/pkg/sql/logictest/testdata/logic_test/geospatial) for concatenating array components together.

@otan

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=82304) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=82304) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=82304) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82304)
<!-- Reviewable:end -->

otan reviewed (changes requested):
see comments about modifying pg_type

You'll also want to add some tests concatting array components together in https://github.com/cockroachdb/cockroach/blob/1a707a1/pkg/sql/logictest/testdata/logic_test/geospatial

surahman commented:
Thank you for all the details!

> Add the relevant logic tests.
> 
> Make sure entries here are updated
> 
> https://github.com/cockroachdb/cockroach/blob/1a707a1f01427f3279dfd425e8b6f2115ceb09da/pkg/sql/logictest/testdata/logic_test/pg_catalog#L1958
> 
> You can use
> 
> ```
> make testbaselogic FILES='pg_catalog' TESTFLAGS='-rewrite'
> ```
> 
> To automatically rewrite the data

Is there a way for me to run this command from Bazel? I have not gone the old `make` route for building because my understanding is it will be deprecated. The project is not rooted on `$GOPATH` and is imported into Goland, with some issues, using the Bazel plugin. Is the preferred way to setup the project to root on the `$GOPATH`?

It would be convenient to have setup instructions for the Goland IDE for engagement with developers. At the moment I am running into issues with being able to run the individual tests from within the IDE using the column icons. This means I cannot debug from within the tests.

> Please also fold your commits into one commit!

I will squash all the commits on the branch into a single commit. I am keeping the commit tree in the event I make a mistake and need to drop/revert commits during development.

> You'll also want to add some tests concatting array components together in https://github.com/cockroachdb/cockroach/blob/1a707a1/pkg/sql/logictest/testdata/logic_test/geospatial

I will look into getting these tests created.

surahman commented on pkg/sql/types/types.go:
```diff
@@ -2812,3 +2812,15 @@ func (m *GeoMetadata) SQLString() string {
 }
+
+// Delimiter selects the correct delimiter rune based on the datum type specified.
+func Delimiter(t *T) string {
```
Great point - change affected.

surahman commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -4438,7 +4438,7 @@ func (d *DArray) Format(ctx *FmtCtx) {
 		ctx.WriteString(comma)
 		ctx.FormatNode(v)
-		comma = ","
+		comma = types.Delimiter(d.ParamTyp)
```
Thank you, change affected.

otan commented on pkg/sql/logictest/testdata/logic_test/geospatial_bbox:
```diff
@@ -20,7 +20,7 @@ query ITTT
 ----
 1  BOX(1 2,3 4)              BOX(3 4,5 6)      {"BOX(-1 -2,-3 -4)"}
-2  BOX(10.1 20.1,30.5 40.6)  BOX(30 40,50 60)  {"BOX(-1 -2,-3 -4)","BOX(3 -4,5 -6)"}
+2  BOX(10.1 20.1,30.5 40.6)  BOX(30 40,50 60)  {"BOX(-1 -2,-3 -4)";"BOX(3 -4,5 -6)"}
```
this should still be a `,`

otan commented on pkg/sql/logictest/testdata/logic_test/geospatial:
```diff
@@ -159,7 +159,7 @@ INSERT INTO geo_array_table VALUES (
 SELECT * FROM geo_array_table
 ----
-1  {0101000020E6100000000000000000F03F000000000000F03F,0102000020E6100000020000000000000000000040000000000000004000000000000008400000000000000840}  {0101000000000000000000F03F000000000000F03F,0102000000020000000000000000000040000000000000004000000000000008400000000000000840}
+1  {0101000020E6100000000000000000F03F000000000000F03F,0102000020E6100000020000000000000000000040000000000000004000000000000008400000000000000840}  {0101000000000000000000F03F000000000000F03:0102000000020000000000000000000040000000000000004000000000000008400000000000000840}
```
why is LHS `,` and RHS `:`? they should both use `:`

otan commented on pkg/sql/types/types.go:
```diff
@@ -2812,3 +2812,15 @@ func (m *GeoMetadata) SQLString() string {
+	switch t.Family() {
+	case Geometry.Family(), Geography.Family():
+		return ":"
+	case Box2D.Family():
```
this is my bad; i think box2d is still `,` - you can remove this line.

otan reviewed (commented):
the tests are going to fail to a parse error - see below comment

otan reviewed (commented):
you'll also need to modify

https://github.com/cockroachdb/cockroach/blob/dcfe20c25651494692a5b1132a72d5702242f394/pkg/sql/sem/tree/parse_array.go#L192

to take the delimiter into effect

surahman commented:
Should we also be modifying the `(d *DTuple) pgwireFormat(ctx *FmtCtx)` to handle delimiters based on actual types or is this strictly for arrays?

https://github.com/cockroachdb/cockroach/blob/aa56262a5e0457e9e56304ade5ca2b7673c8f75f/pkg/sql/sem/tree/pgwire_encode.go#L36

otan commented:
only arrays

surahman commented:
It seems as though the two errors below are related to how the datum string is parsed into a `tree.Datum`:

<details>
<summary>TestRandParseDatumStringAs</summary>

```bash
=== RUN   TestRandParseDatumStringAs/geography[]
    roundtrip_format_test.go:101: ARRAY['01030000A0E6100000010000000D000000E4265B7CB33556C000646C250C31EE3F90F5DC4813B1E441808EE0481F001340002B842AA2E72F40189AE7E68698EF41A8D091613DE335C040DCF8569EDE1B40AC9440A60576F041F8A1903FD4A660C0DA3AB7C573EB53C018DC14386D33DF41B85AD222FCE547402030D72C8D173FC0BA020818D64CF8C11A948BEA9F066540B081D0E997772640E4467C8CEF50F1416E67876B27086640903C2EED00F63B40CF7EB1DEA26EF4C1E489546AC4915F40CCF1BA4977EC4440FCA3BE22872AF941DC5EC32251706340FA3D60A7045152404CA40C283AF1E541E4577E3BBEB65C40A8B050BA5E3654402D84642A9E1EF4C1F07C2481C230554006CC9010D121544000E4E08BE49C79C1645E1A0D805661C00E5EBAD771CE53406A22426F3C59E0C1E4265B7CB33556C000646C250C31EE3F90F5DC4813B1E441':::GEOGRAPHY:'01020000E0E61000000200000057CAFB07229053C000A32A0C649E0F409218B68BDB02E9C17E1D9DC22B56F0416812A4B33875324070ADFB174E7435408C3998E3AF65F1C197BBDD261F1C01C2':::GEOGRAPHY:'0103000020E6100000010000000500000000000000000000000000000000000000000000000000F03F0000000000000000000000000000F03F000000000000F03F00000000000
00000000000000000F03F00000000000000000000000000000000':::GEOGRAPHY] at or near ":": syntax error
```

```bash
=== RUN   TestRandParseDatumStringAs/geometry[]
    roundtrip_format_test.go:101: ARRAY['01030000C000000000':::GEOMETRY:'01010000C0B8FF4ED1A3E9DC41F04B669AF47200C2242C8449FA4FE741A0D932B0426D00C2':::GEOMETRY] at or near ":": syntax error

```

</details>

It seems as though it is the case that the new delimiter `:` for `geography` and `geometry` is not parsed correctly. The `ParseDatumStringAs` seems to be the cause of the issue. I am unable to run the test suite and attach the debugger from within Goland to run the tests and track this down.
https://github.com/cockroachdb/cockroach/blob/d8d995dfe7a50e9a66a3b1e45c867ca8592986f3/pkg/sql/rowenc/roundtrip_format.go#L26-L27

otan commented:
> Judging from the error message I suspect the issue with the Avro test is also related to parsing.

I think the `ARRAY[..]` syntax still uses `,`

surahman commented:
I am still getting familiar with the code base and getting my system setup but the error messages do show a `:` between the `ARRAY` elements

```bash
ARRAY['01030000C000000000':::GEOMETRY:'010100 ...
____________________________________/^
```

otan commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -4434,11 +4434,11 @@ func (d *DArray) Format(ctx *FmtCtx) {
+		ctx.WriteString(delimiter)
 		ctx.FormatNode(v)
-		comma = ","
+		delimiter = d.ParamTyp.Delimiter()
```
The error is here - we should undo this. I believe the change only affects the {} syntax.

surahman replied:
I will drop that commit.

surahman replied:
For posterity: I removed the changes to line 4441:

```golang
delimiter = d.ParamTyp.Delimiter()
```

otan replied:
hmm, this still is here and causing the tests to fail. i think all of this change should be removed.

surahman replied:
My bad I typoed a Git command to drop the commit.

otan commented:
also, please squash all your changes into one commit! thanks.

surahman committed:
[CRDB-15892] SQL delimiter for Geometry and Geography inconsistent.

Release note
The delimiters for Geography and Geometry of PostgreSQL is inconsistent. It has been updated to use the ":" delimiter.

otan reviewed (approved):
thanks for persisting with this and me on this! this was definitely trickier than i expected.

bors r+

surahman commented:
Great working with you on this. Thank you for all your help, patience, and guidance. `RoachDB is a very interesting piece of technology.

andreimatei commented:
I believe this patch made `TestLogic/local/pg_catalog` flaky, as indicated by
```
./dev test --stress pkg/sql/logictest -f=TestLogic/local/pg_catalog
```
in anywhere from 1 to 100 runs.
I've tested before and after.

The type delimiter seems to sometimes change in lines like (see comma vs colon):
```
                    4294967004  spatial_ref_sys                        C            false           true          ,         4294967004  0        0
```
to:
```
                    4294967004  spatial_ref_sys                        C            false           true          :         4294967004  0        0
```


@otan will you kindly look into it?

rafiss commented:
I looked briefly. the bug is that a global variable is assigned here, when it shouldn't be: https://github.com/cockroachdb/cockroach/blob/02cc3937046d4b5642e715b0fce77e4037e21d61/pkg/sql/pg_catalog.go#L2905

i can fix
